File: realtime_analysis.m
clc, clear, close all
lw = 2;

t4 = load("time_test4.mat");
t5 = load("time_test5.mat");
t10 = load("time_test10.mat");
t11 = load("time_test11.mat");
t12 = load("time_test12.mat");
t13 = load("time_test13.mat");
t14 = load("time_test14.mat");
t15 = load("time_test15.mat");
t16 = load("time_test16.mat");

t12sp = load("time_test12sp.mat");





t4 = t4.timer;
t5 = t5.timer;
t10 = t10.timer;
t11 = t11.timer;
t12 = t12.timer;
t13 = t13.timer;
t14 = t14.timer;
t15 = t15.timer;
t16 = t16.timer;
t12sp = t12sp.timer;

figure
plot(t12,'o-','LineWidth', lw), hold on
plot(t12sp,'o-','LineWidth', lw), hold on
plot(t13,'o-','LineWidth', lw)
plot(t14,'o-','LineWidth', lw)
grid on
legend("No objects (scenario no.0)", "1 Object (scenario no.1)", "3 Objects (scenario no.2)",...
    "12 Objects (scenario no.3)")
xlabel("Elapsed simulation time (s)")
ylabel("Computed time (s)")
title("Various scenarios with dynamic environmental constraints")
set(gca, 'FontSize', 24, 'LineWidth', 2)


figure
plot(t4,'o-','LineWidth', lw), hold on
plot(t10,'o-','LineWidth', lw)
plot(t15,'o-','LineWidth', lw)
grid on
legend("No environmental constraints", "Static environmental constraints",...
    "Dynamic environmental constraints")
xlabel("Elapsed simulation time (s)")
ylabel("Computed time (s)")
title("Scenario no.4 : 1 static obstacle + 2 dynamic obstacles")
set(gca, 'FontSize', 24, 'LineWidth', 2)


figure
plot(t5,'o-','LineWidth', lw), hold on
plot(t11,'o-','LineWidth', lw)
plot(t16,'o-','LineWidth', lw)
grid on
legend("No environmental constraints", "Static environmental constraints",...
    "Dynamic environmental constraints")
xlabel("Elapsed simulation time (s)")
ylabel("Computed time (s)")
title("Scenario no.5 : 2 static obstacles + 2 dynamic obstacles")
set(gca, 'FontSize', 24, 'LineWidth', 2)

File: scripts/realtime_analysis.py
"""Port of ``real-time_analysis/realtime_analysis.m``.

Plots stored ``timer`` data from previous simulation runs to compare compute
time across scenarios and environment conditions.
"""

from __future__ import annotations

import argparse
from pathlib import Path

import matplotlib.pyplot as plt
from scipy.io import loadmat

REPO_ROOT = Path(__file__).resolve().parents[1]
RTA_DIR = REPO_ROOT / "data" / "real-time_analysis"


def _load_timer(name: str) -> dict | None:
    path = RTA_DIR / name
    if not path.exists():
        print(f"[warn] {path} not found")
        return None
    d = loadmat(str(path))
    return d


def run() -> None:
    names = {
        "t4": "time_test4.mat", "t5": "time_test5.mat",
        "t10": "time_test10.mat", "t11": "time_test11.mat",
        "t12": "time_test12.mat", "t13": "time_test13.mat",
        "t14": "time_test14.mat", "t15": "time_test15.mat",
        "t16": "time_test16.mat", "t12sp": "time_test12sp.mat",
    }
    timers = {k: d["timer"].ravel() for k, v in names.items() if (d := _load_timer(v)) is not None}

    lw = 2

    if all(k in timers for k in ("t12", "t12sp", "t13", "t14")):
        fig, ax = plt.subplots()
        for k, label in zip(
            ("t12", "t12sp", "t13", "t14"),
            ("No objects", "1 Object", "3 Objects", "12 Objects"),
        ):
            ax.plot(timers[k], "o-", linewidth=lw, label=label)
        ax.set_xlabel("Elapsed simulation time (s)")
        ax.set_ylabel("Computed time (s)")
        ax.set_title("Various scenarios with dynamic environmental constraints")
        ax.legend(); ax.grid(True)

    if all(k in timers for k in ("t4", "t10", "t15")):
        fig, ax = plt.subplots()
        for k, label in zip(
            ("t4", "t10", "t15"),
            ("No env. constraints", "Static env. constraints", "Dynamic env. constraints"),
        ):
            ax.plot(timers[k], "o-", linewidth=lw, label=label)
        ax.set_xlabel("Elapsed simulation time (s)")
        ax.set_ylabel("Computed time (s)")
        ax.set_title("1 static + 2 dynamic obstacles")
        ax.legend(); ax.grid(True)

    if all(k in timers for k in ("t5", "t11", "t16")):
        fig, ax = plt.subplots()
        for k, label in zip(
            ("t5", "t11", "t16"),
            ("No env. constraints", "Static env. constraints", "Dynamic env. constraints"),
        ):
            ax.plot(timers[k], "o-", linewidth=lw, label=label)
        ax.set_xlabel("Elapsed simulation time (s)")
        ax.set_ylabel("Computed time (s)")
        ax.set_title("2 static + 2 dynamic obstacles")
        ax.legend(); ax.grid(True)

    plt.show()


def main() -> None:
    argparse.ArgumentParser(description="Realtime analysis plots").parse_args()
    run()


if __name__ == "__main__":
    main()
